Alcatel-Lucent Enterprise AOS8 integration

jefvantongerloo opened the pull request:
## Test inclusion requirements

In case the PR contains an enhancement or a new platform/service support, some tests have to be added for the new functionality:

- any fix or enhancement SHOULD include relevant new tests or test updates, if any tests need updating.
- a new platform support MUST include the relevant input files similar to what we have in tests/integration/sqcmds/-input directories, along with the relevant tests in the tests/integration/sqcmds/-samples dir. That list MUST include the all.yml file fully filled out.
- any new service (or table) addition MUST include comments about what network OS are supported (along with version) with this command along with test samples for those platforms and input files in the *-input dir

For additional information about tests, follow this [link](https://suzieq.readthedocs.io/en/latest/developer/testing/)

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- New feature (non-breaking change which adds functionality)

## Changes to the Documentation

<!--
If the docs must be updated, please include the changes in the PR.
-->

...

## Proposed Release Note Entry

<!--
Please provide a short summary of your PR that we can copy & paste
into the release notes.
-->

...

## Comments


## Double Check

<!--
Please put an x into the brackets (like `[x]`) if you've completed that task.
-->

- [x] I have read the comments and followed the [CONTRIBUTING.md](https://github.com/netenglabs/suzieq/blob/master/CONTRIBUTING.md).
- [x] I have explained my PR according to the information in the comments or in a linked issue.
- [x ] My PR source branch is created from the `develop` branch.
- [x ] My PR targets the `develop` branch.
- [x] All my commits have `--signoff` applied

claudious96 commented on suzieq/poller/worker/services/interfaces.py:
```diff
@@ -65,6 +65,65 @@ def _textfsm_valid_speed_value(self, entry):
+            mtu = interfaces.get('long_frame_size', -1)
+            numChanges = interfaces.get('number_of_status_change', -1)
+            reason = interfaces.get('port_down_violation_reason', -1)
+            speed = interfaces_status.get('detected_speed', -1)
```
Instead of using hardcoded -1, use the `_textfsm_valid_speed_value` method to automatically handle the value and default value in case of missing speed. Just be sure to:
- have the key for `type` in the dictionary passed to the func
- store the speed value in the dictionary as 'speed' instead of 'detected_speed'  and to return an empty string as default value.

jefvantongerloo replied:
Adjusted to use the function, but gives the same results as before.

claudious96 commented on suzieq/poller/worker/services/macs.py:
```diff
@@ -52,6 +52,40 @@ def _add_mackey_protocol(self, entry):
+            vlan = entry.get('id', -1)
+            macaddr = entry.get('mac_address', -1)
+            #Where does this parameter go?
+            entry.get('type', -1)
```
I believe this is a leftover, we don't have a `type` value field in the macs table. I think there was a discussion on slack about placing this piece information in the 'flags` field but I'm not sure whether that was cleared out.

jefvantongerloo replied:
Adjusted to use the function, but gives the same results as before.

jefvantongerloo replied:
Removed `type` field.

claudious96 commented on suzieq/poller/worker/services/macs.py:
```diff
@@ -52,6 +52,40 @@ def _add_mackey_protocol(self, entry):
+                            'macaddr': macaddr,
+                            'flags': flags,
+                            'oif': oif,
+                            'type': type
```
Same as before, we don't have a `type` field in the macs table.

jefvantongerloo replied:
Removed `type` field.

claudious96 commented on suzieq/poller/worker/services/routes.py:
```diff
@@ -36,6 +36,22 @@ def _common_data_cleaner(self, processed_data, _):
+                'prefix': entry['dest_addr'],
+                'nexthopIps': [entry['gateway_addr']],
+                'protocol': entry['protocol'],
+            }
```
Would it be possible to retrieve also fields like: `oifs`, `source`, `vrf`, `action`?

jefvantongerloo replied:
`oifs` could be added, but another more complex command has to be parsed: `show ip router database`

claudious96 commented on suzieq/poller/worker/services/interfaces.py:
```diff
@@ -65,6 +65,65 @@ def _textfsm_valid_speed_value(self, entry):
+            else:
+                adminState = "down"
+
+            entry_dict = {
+                'adminState': adminState,
+                'ifname': ifname,
+                'interfaceMac': interfaceMac,
+                'mtu': mtu,
+                'numChanges': numChanges,
+                'reason': reason,
+                'speed': speed,
+                'state': state,
+                'type': type,
+                'vlan': vlan,
+                'vlanList' : vlanList
+            }
```
Would it be possible to retrieve also `ipAddressList` and `master` (vrf) fields?

claudious96 commented on suzieq/poller/worker/services/arpnd.py:
```diff
@@ -10,6 +10,16 @@
+        entry['macaddr'] = entry.get('hardware_addr', '-1')
+        entry['oif'] = entry.get('port', '-1') # or 'port' for physical interface
+
+        return processed_data
```
Do you think it could be possible to retrieve also the `state` field (reachable, permanent .. )?

jefvantongerloo replied:
Added ´state´ column from ´type´ field:
STATIC = permanent
DYNAMIC = reachable

claudious96 reviewed (changes requested):
First of all thanks for the work!

I left some minor comments about styling or linting-related. You can see the reason in the github actions for flake8 and pylint, or just run locally `pylint suzieq` and `flake8 suzieq`.

Others are related to default values in the cleanup functions in services and about some missing fields for tables entries: arpnd, interfaces, routes.

jefvantongerloo commented on suzieq/poller/worker/nodes/node.py:
```diff
@@ -1277,6 +1285,45 @@ def _extract_nos_version(self, data) -> None:
+            data = output[0]['data']
+
+            # Extract hostname
+            hname = re.search('\s+Name:\s+(\S+),', data)
```
Added raw string literal for all regex functions.

jefvantongerloo commented:
Forgot to signoff latest commits and rebased with --signoff parameter.

jefvantongerloo commented on suzieq/poller/worker/nodes/node.py:
```diff
@@ -494,7 +499,8 @@ async def _get_device_type_hostname(self):
         # hostnamectl on Linux systems. That's all we support today.
         await self._exec_cmd(self._parse_device_type_hostname,
-                             ["show version",
+                             ["show system",
+                              "show version",
```
This really should be fixed with a durable solution.

There is a problem with the ´only_one´ parameters on the cmd execution in function ´_get_device_type_hostname´ for the device discovery. If an unknown command is sent to the device, status 0 and an error message in standard output get returned. So when the correct command for that type of device is not used first, the device gets discovered as unknown.

The program can also never enter the elif function because only one command from the list is sent if the return status is 0 (even with an error return):

```
        elif (len(output) > 1) and (output[1]["status"] == 0):
            devtype = 'unsupported'
            data = output[1]["data"]
            if data:
                if "Cumulus Linux" in data:
                    devtype = "cumulus"
                else:
                    devtype = "linux"

                version_str = data
                # Hostname is the last line of the output
                if len(data.strip()) > 0:
                    hostname = data.splitlines()[-1].strip()
```

Sample terminal output on unknown cli command (with return status code 0):

```
net-swi-001> show version
                     ^
ERROR: Invalid entry: "version"
```

claudious96 replied:
> This really should be fixed with a durable solution.

I agree. What you propose, checking for a command failure using both the return code AND a regex for the output, seems the best way to go.

For this first phase I'd go for supporting Alcatel-Lucent only supplying `devtype` as `aos` in the inventory, we can implement the new way of determining a failed command separately and then come back to having the autodiscovery.

> Sample terminal output on unknown cli command (with return status code 0):
> 
> net-swi-001> show version
>                      ^
> ERROR: Invalid entry: "version"

Does this appear in the stdout or stderr?

jefvantongerloo replied:
Terminal output from cli command, so stdout.

jefvantongerloo commented:
- Updated and tested the textfsm templates in latest aos releases (8.9.73.R01 & 8.8.56.R02).
- Added ip interfaces en aggregate interfaces
- Added ip and vlan members on route and interface services
- Added oif on router service
